pootle/apps/pootle_store/models.py

```diff
@@ -1867,143 +1737,82 @@ def update(self, overwrite=False, store=None, only_newer=False, user=None,
             logging.info(u"Attempted to update %s while locked",
                          self.pootle_path)
             return
-        elif self.state < PARSED:
-            # File has not been parsed before
-            logging.debug(u"Attempted to update unparsed file %s",
-                          self.pootle_path)
-            self.parse(store=store)
-            return
-
-        disk_mtime = self.get_file_mtime()
-        if only_newer and disk_mtime == self.file_mtime:
-            # The file on disk wasn't changed since the last sync
-            logging.debug(u"File didn't change since last sync, skipping "
-                          u"%s" % self.pootle_path)
-            return
-
-        if store is None:
-            store = self.file.store
 
         # Lock store
         logging.debug(u"Updating %s", self.pootle_path)
         old_state = self.state
         self.state = LOCKED
-        self.save()
+        self.save(update_cache=False)
 
-        store_revision = parse_pootle_revision(store)
-        max_unit_revision = self.get_max_unit_revision()
-        update_revision = Revision.incr()
-        try:
-            changes = {
-                'obsolete': 0,
-                'updated': 0,
-                'added': 0,
-            }
-
-            old_unitid_list = []  # unitid list for comparing (DB)
-            old_unitids = {}  # unitids dict (DB)
-            old_unitid_set = set()
-            old_obsolete_unitid_set = set()  # obsolete unitids in DB
-            update_unitids = {}
+        if user is None:
+            User = get_user_model()
+            user = User.objects.get_system_user()
 
-            unit_values = self.unit_set.order_by('index').values_list(
-                'unitid', 'state', 'id', 'index',
-            )
-            for (unitid, state, dbid, index) in unit_values:
-                if state == OBSOLETE:
-                    old_obsolete_unitid_set.add(unitid)
+        update_revision = None
+        changes = {}
+        try:
+            to_change = StoreDiff(self, store, store_revision).diff()
+            change = False
```

Comment:
What about having the logic to determine if a store has changes in a static method of `StoreDiff`? The method can be tested on its own, and this becomes easier to read (I'm pretty happy with what you have already achieved in `Store.update()`, this would make it a breeze to look at though):
```python
to_change = StoreDiff(self, store, store_revision).diff()
if StoreDiff.has_changes(to_change):
     ...
```